dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBMongoDB/AzureCosmosDBMongoDBGenericDataModelMapper.cs

```diff
@@ -0,0 +1,181 @@
+                {
+                    continue;
+                }
+
+                vectorProperties.Add(vectorProperty.DataModelPropertyName, GetVectorPropertyValue(property.PropertyType, vectorValue));
+            }
+        }
+
+        if (key is null)
+        {
+            throw new VectorStoreRecordMappingException("No key property was found in the record retrieved from storage.");
+        }
+
+        return new VectorStoreGenericDataModel<string>(key) { Data = dataProperties, Vectors = vectorProperties };
+    }
+
+    #region private
+
+    private static object? GetDataPropertyValue(Type propertyType, BsonValue value)
+    {
+        if (value.IsBsonNull)
+        {
+            return null;
+        }
+
+        return propertyType switch
+        {
+            Type t when t == typeof(bool) => value.AsBoolean,
+            Type t when t == typeof(bool?) => value.AsNullableBoolean,
+            Type t when t == typeof(string) => value.AsString,
+            Type t when t == typeof(int) => value.AsInt32,
+            Type t when t == typeof(int?) => value.AsNullableInt32,
+            Type t when t == typeof(long) => value.AsInt64,
+            Type t when t == typeof(long?) => value.AsNullableInt64,
+            Type t when t == typeof(float) => ((float)value.AsDouble),
+            Type t when t == typeof(float?) => ((float?)value.AsNullableDouble),
+            Type t when t == typeof(double) => value.AsDouble,
+            Type t when t == typeof(double?) => value.AsNullableDouble,
+            Type t when t == typeof(decimal) => value.AsDecimal,
+            Type t when t == typeof(decimal?) => value.AsNullableDecimal,
+            Type t when t == typeof(DateTime) => value.ToUniversalTime(),
+            Type t when t == typeof(DateTime?) => value.ToNullableUniversalTime(),
+            Type t when typeof(IEnumerable).IsAssignableFrom(t) => value.AsBsonArray.Select(
+                item => GetDataPropertyValue(VectorStoreRecordPropertyReader.GetCollectionElementType(t), item)),
+            _ => throw new NotSupportedException($"Mapping for type {propertyType.FullName} is not supported in generic data model.")
+        };
+    }
+
+    private static object? GetVectorPropertyValue(Type propertyType, BsonValue value)
+    {
+        if (value.IsBsonNull)
+        {
+            return null;
+        }
+
+        return propertyType switch
+        {
+            Type t when t == typeof(ReadOnlyMemory<float>) || t == typeof(ReadOnlyMemory<float>?) =>
+                new ReadOnlyMemory<float>(value.AsBsonArray.Select(item => (float)item.AsDouble).ToArray()),
+            Type t when t == typeof(ReadOnlyMemory<double>) || t == typeof(ReadOnlyMemory<double>?) =>
+                new ReadOnlyMemory<double>(value.AsBsonArray.Select(item => item.AsDouble).ToArray()),
+            _ => throw new NotSupportedException($"Mapping for type {propertyType.FullName} is not supported in generic data model.")
```

Comment:
It would be nice to include the property that the failure happened on in the error message here and in the other cases too since the user isn't dealing with a class that they can easily see their property types on but rather an object that is probably built dynamically, making it harder to troubleshoot.

Reply:
Good idea, added!